File: sdk/python/src/flare_redact/wire.py
"""Translation between Python options and the gateway's JSON wire format.

The gateway speaks the JavaScript library's option names (camelCase). Keeping the
translation in one small module means the Python API can stay Pythonic without
the two drifting.

One deliberate omission: ``transform_secret`` is never serialised. A keyed
transform's whole value is that the key stays where it was configured; shipping
it to a sidecar on every request would defeat that. Configure the secret on the
gateway instead.
"""

from __future__ import annotations

import re
from typing import Any, Dict, Mapping, Union

from .engine import Finding, Options

__all__ = [
    "options_to_wire", "options_from_wire", "finding_from_wire", "finding_to_wire",
    "UnserializableOptionError",
]

_SIMPLE_FIELDS = (
    ("only", "only"),
    ("enable", "enable"),
    ("disable", "disable"),
    ("mode", "mode"),
    ("terms", "terms"),
    ("terms_case_sensitive", "termsCaseSensitive"),
    ("min_confidence", "minConfidence"),
    ("refine_confidence", "refineConfidence"),
    ("include_values", "includeValues"),
)

_DEFAULTS = Options()


class UnserializableOptionError(TypeError):
    """Raised when an option cannot cross a process boundary."""


def options_to_wire(options: Union[Options, Mapping[str, Any], None]) -> Dict[str, Any]:
    """Serialise options for ``POST /v1/redact``, omitting anything at its default."""
    resolved = Options.coerce(options)
    wire: Dict[str, Any] = {}

    for python_name, wire_name in _SIMPLE_FIELDS:
        value = getattr(resolved, python_name)
        if value != getattr(_DEFAULTS, python_name):
            wire[wire_name] = list(value) if isinstance(value, tuple) else value

    if isinstance(resolved.mask, str):
        wire["mask"] = resolved.mask
    elif callable(resolved.mask):
        raise UnserializableOptionError(
            "A callable mask cannot be sent to the gateway; redact locally or use a mask string."
        )

    if resolved.redact_keys is not True:
        if isinstance(resolved.redact_keys, re.Pattern):
            raise UnserializableOptionError(
                "A compiled redact_keys pattern cannot be sent to the gateway; pass an exact list."
            )
        wire["redactKeys"] = (
            list(resolved.redact_keys) if not isinstance(resolved.redact_keys, bool) else resolved.redact_keys
        )

    if resolved.allow is not None:
        if isinstance(resolved.allow, re.Pattern):
            raise UnserializableOptionError(
                "A compiled allow pattern cannot be sent to the gateway; pass an exact list."
            )
        wire["allow"] = list(resolved.allow)

    limits: Dict[str, int] = {}
    if resolved.max_input_length != _DEFAULTS.max_input_length:
        limits["maxInputLength"] = resolved.max_input_length
    if resolved.max_findings != _DEFAULTS.max_findings:
        limits["maxFindings"] = resolved.max_findings
    if limits:
        wire["limits"] = limits

    return wire


def options_from_wire(payload: Mapping[str, Any]) -> Options:
    """Build options from the JSON wire format used by the gateway and the corpus.

    Unlike :func:`options_to_wire`, this *does* accept ``transformSecret``: a
    gateway may be configured with one, and the conformance corpus needs fixed
    keyed-transform vectors.
    """
    limits = payload.get("limits") or {}
    known = {
        "only", "enable", "disable", "mode", "mask", "transformSecret", "redactKeys",
        "allow", "terms", "termsCaseSensitive", "minConfidence", "refineConfidence",
        "includeValues", "limits",
    }
    unknown = set(payload) - known
    if unknown:
        raise ValueError(f"unknown options: {', '.join(sorted(unknown))}")

    defaults = Options()
    return Options(
        only=payload.get("only"),
        enable=payload.get("enable"),
        disable=payload.get("disable"),
        mode=payload.get("mode", defaults.mode),
        mask=payload.get("mask"),
        transform_secret=payload.get("transformSecret"),
        redact_keys=payload.get("redactKeys", defaults.redact_keys),
        allow=payload.get("allow"),
        terms=payload.get("terms"),
        terms_case_sensitive=bool(payload.get("termsCaseSensitive", defaults.terms_case_sensitive)),
        min_confidence=float(payload.get("minConfidence", defaults.min_confidence)),
        refine_confidence=bool(payload.get("refineConfidence", defaults.refine_confidence)),
        include_values=bool(payload.get("includeValues", defaults.include_values)),
        max_input_length=int(limits.get("maxInputLength", defaults.max_input_length)),
        max_findings=int(limits.get("maxFindings", defaults.max_findings)),
    )


def finding_to_wire(finding: Finding) -> Dict[str, Any]:
    """Serialise a finding using the wire field names."""
    return finding.to_dict()


def finding_from_wire(payload: Mapping[str, Any]) -> Finding:
    """Rebuild a :class:`Finding` from the gateway's JSON representation."""
    return Finding(
        detector=str(payload.get("detector", "")),
        label=str(payload.get("label", "")),
        why=str(payload.get("why", "")),
        risk=str(payload.get("risk", "high")),
        confidence=float(payload.get("confidence", 0.0)),
        start=payload.get("start"),
        end=payload.get("end"),
        line=payload.get("line"),
        column=payload.get("column"),
        path=payload.get("path"),
        value=payload.get("value"),
    )

File: sdk/go/wire.go
package flareredact

import (
	"encoding/json"
	"fmt"
	"sort"
)

// Translation between Go options and the JSON wire format used by the gateway
// and the conformance corpus.
//
// The wire format is the JavaScript library's option set (camelCase), so one
// document configures every engine. Keeping the mapping in one small file means
// the Go API can stay idiomatic without the two drifting.

var wireOptionNames = map[string]struct{}{
	"only": {}, "enable": {}, "disable": {}, "mode": {}, "mask": {},
	"transformSecret": {}, "redactKeys": {}, "allow": {}, "terms": {},
	"termsCaseSensitive": {}, "minConfidence": {}, "refineConfidence": {},
	"includeValues": {}, "limits": {},
}

func wireStrings(value any, field string) ([]string, error) {
	items, ok := value.([]any)
	if !ok {
		return nil, fmt.Errorf("flareredact: %q must be an array of strings", field)
	}
	out := make([]string, 0, len(items))
	for _, item := range items {
		text, ok := item.(string)
		if !ok {
			return nil, fmt.Errorf("flareredact: %q must be an array of strings", field)
		}
		out = append(out, text)
	}
	return out, nil
}

func wireTerms(value any) ([]Term, error) {
	switch typed := value.(type) {
	case []any:
		out := make([]Term, 0, len(typed))
		for _, item := range typed {
			switch entry := item.(type) {
			case string:
				out = append(out, Term{Term: entry})
			case map[string]any:
				term, _ := entry["term"].(string)
				replace, _ := entry["replace"].(string)
				out = append(out, Term{Term: term, Replace: replace})
			default:
				return nil, fmt.Errorf("flareredact: %q entries must be strings or objects", "terms")
			}
		}
		return out, nil
	case map[string]any:
		// A {term: replacement} object. Sort so the detector alternation is
		// built in the same order on every run.
		keys := make([]string, 0, len(typed))
		for key := range typed {
			keys = append(keys, key)
		}
		sort.Strings(keys)
		out := make([]Term, 0, len(keys))
		for _, key := range keys {
			replace, _ := typed[key].(string)
			out = append(out, Term{Term: key, Replace: replace})
		}
		return out, nil
	}
	return nil, fmt.Errorf("flareredact: %q must be an array or an object", "terms")
}

// OptionsFromWire builds Options from the JSON wire representation.
func OptionsFromWire(payload map[string]any) (Options, error) {
	var options Options
	for key := range payload {
		if _, ok := wireOptionNames[key]; !ok {
			return options, fmt.Errorf("flareredact: unknown option %q", key)
		}
	}

	var err error
	if value, ok := payload["only"]; ok {
		if options.Only, err = wireStrings(value, "only"); err != nil {
			return options, err
		}
	}
	if value, ok := payload["enable"]; ok {
		if options.Enable, err = wireStrings(value, "enable"); err != nil {
			return options, err
		}
	}
	if value, ok := payload["disable"]; ok {
		if options.Disable, err = wireStrings(value, "disable"); err != nil {
			return options, err
		}
	}
	if value, ok := payload["allow"]; ok {
		if options.Allow, err = wireStrings(value, "allow"); err != nil {
			return options, err
		}
	}
	if value, ok := payload["mode"].(string); ok {
		options.Mode = Mode(value)
	}
	if value, ok := payload["mask"].(string); ok {
		options.Mask = value
	}
	if value, ok := payload["transformSecret"].(string); ok {
		options.TransformSecret = value
	}
	if value, ok := payload["redactKeys"]; ok {
		switch typed := value.(type) {
		case bool:
			options.DisableKeyRedaction = !typed
		default:
			if options.KeyNames, err = wireStrings(value, "redactKeys"); err != nil {
				return options, err
			}
		}
	}
	if value, ok := payload["terms"]; ok {
		if options.Terms, err = wireTerms(value); err != nil {
			return options, err
		}
	}
	if value, ok := payload["termsCaseSensitive"].(bool); ok {
		options.TermsCaseSensitive = value
	}
	if value, ok := payload["minConfidence"].(float64); ok {
		options.MinConfidence = value
	}
	if value, ok := payload["refineConfidence"].(bool); ok {
		options.RefineConfidence = value
	}
	if value, ok := payload["includeValues"].(bool); ok {
		options.IncludeValues = value
	}
	if limits, ok := payload["limits"].(map[string]any); ok {
		if value, ok := limits["maxInputLength"].(float64); ok {
			options.MaxInputLength = int(value)
		}
		if value, ok := limits["maxFindings"].(float64); ok {
			options.MaxFindings = int(value)
		}
	}
	return options, nil
}

// OptionsToWire serialises the options a gateway will accept, omitting defaults.
//
// TransformSecret is deliberately not included: a keyed transform's whole value
// is that the key stays where it was configured, so it is set on the gateway
// rather than shipped with every request.
func OptionsToWire(options Options) map[string]any {
	wire := map[string]any{}
	if len(options.Only) > 0 {
		wire["only"] = options.Only
	}
	if len(options.Enable) > 0 {
		wire["enable"] = options.Enable
	}
	if len(options.Disable) > 0 {
		wire["disable"] = options.Disable
	}
	if len(options.Allow) > 0 {
		wire["allow"] = options.Allow
	}
	if options.Mode != "" && options.Mode != ModeMask {
		wire["mode"] = string(options.Mode)
	}
	if options.Mask != "" {
		wire["mask"] = options.Mask
	}
	if options.DisableKeyRedaction {
		wire["redactKeys"] = false
	} else if len(options.KeyNames) > 0 {
		wire["redactKeys"] = options.KeyNames
	}
	if len(options.Terms) > 0 {
		terms := make([]any, 0, len(options.Terms))
		for _, term := range options.Terms {
			if term.Replace == "" {
				terms = append(terms, term.Term)
				continue
			}
			terms = append(terms, map[string]any{"term": term.Term, "replace": term.Replace})
		}
		wire["terms"] = terms
	}
	if options.TermsCaseSensitive {
		wire["termsCaseSensitive"] = true
	}
	if options.MinConfidence > 0 {
		wire["minConfidence"] = options.MinConfidence
	}
	if options.RefineConfidence {
		wire["refineConfidence"] = true
	}
	if options.IncludeValues {
		wire["includeValues"] = true
	}
	limits := map[string]any{}
	if options.MaxInputLength > 0 {
		limits["maxInputLength"] = options.MaxInputLength
	}
	if options.MaxFindings > 0 {
		limits["maxFindings"] = options.MaxFindings
	}
	if len(limits) > 0 {
		wire["limits"] = limits
	}
	return wire
}

// MarshalWire is OptionsToWire followed by JSON encoding.
func MarshalWire(options Options) ([]byte, error) { return json.Marshal(OptionsToWire(options)) }
